From af0827f6b48c2d22e644762b63180b354c3817fe Mon Sep 17 00:00:00 2001
From: "Eric S. Raymond" <esr@thyrsus.com>
Date: Mon, 21 Jan 2019 06:26:07 -0500
Subject: [PATCH 5/5] Allow "prefer" on a "pps" peer to use it

This change allows "prefer" on a "pps" peer to indicate that the PPS
input should be used with any source.

The documentation, prior to Richard Lagger's recent cleanup, listed
this as an option, but it did not exist in the code:

"3. If there is a PPS driver and the system clock offset at this point
is less than 0.4 s, and if  there is a prefer peer among the survivors
or if  the PPS  peer is designated  as a prefer  peer, the  PPS driver
becomes the system peer"

Note the option of "if the PPS peer is designated as a prefer peer".

This adds that functionality, which is useful if you want to use the
PPS input with any source. The typical example here would be that you
want to use network sources to number the seconds. Without this
change, you have to mark one as a prefer peer, and the PPS is only
used if that particular peer is selected. This has a couple
problems. First, if that peer goes down, the PPS stops being
used. Second, marking a particular network peer as a prefer peer gives
it special status in the clock selection algorithm, which may not be
desirable.

This resolves GitLab issue #538.  It's a merge by hand of an MR
by Richard Laager.
---
 docs/prefer.adoc | 9 ++++++---
 ntpd/ntp_proto.c | 1 +
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/docs/prefer.adoc b/docs/prefer.adoc
index 13385c50f..e6ea910c0 100644
--- a/docs/prefer.adoc
+++ b/docs/prefer.adoc
@@ -239,9 +239,12 @@ is less than 0.4 s, that PPS driver becomes the system peer and its
 offset and jitter are inherited by the system variables, thus overriding
 any values already computed. However, if the PPS driver is specifically
 the link:driver_pps.html[driver named "pps"], which must be used with
-another driver, then there must be a prefer peer among the survivors, as
-the +prefer+ keyword is used to identify the source associated with the
-PPS input. For an exception, see the +minsane+ option below.
+another driver, additional rules apply. In one mode of operation, there
+must be a prefer peer among the survivors, as the +prefer+ keyword is
+used to identify the source associated with the PPS input. To indicate
+that the PPS input should be used with any source(s), the pps peer
+itself must be marked with the +prefer+ keyword. For an exception, see
+the +minsane+ option below.
 
 If none of the above is the case, the data are disregarded and the
 system variables remain as they are.
diff --git a/ntpd/ntp_proto.c b/ntpd/ntp_proto.c
index 331cf577c..f50a4a4da 100644
--- a/ntpd/ntp_proto.c
+++ b/ntpd/ntp_proto.c
@@ -1925,6 +1925,7 @@ clock_select(void)
 	if (typepps != NULL && fabs(sys_offset) < 0.4 &&
 	    (!typepps->is_pps_driver ||
 	     sys_prefer != NULL ||
+	     (typesystem != NULL && typepps->cfg.flags & FLAG_PREFER) ||
 	     (typesystem == NULL && sys_minsane == 0))) {
 		typesystem = typepps;
 		sys_clockhop = 0;
-- 
2.17.1

